assets/js/util/report-validation.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * Reporting API validation utilities.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Validates data that can be either string or object of the certain type, or array of them.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|string[]|Object|Object[]} data The data to check.
+ * @param {Function} verifyFunction The callback to verify an object.
+ * @return {boolean} TRUE if data is valid, otherwise FALSE.
+ */
+function isValidStringsOrObjects( data, verifyFunction ) {
+	if ( typeof data === 'string' ) {
+		return true;
+	}
+
+	if ( typeof data === 'object' && verifyFunction( data ) ) {
+		return true;
+	}
+
+	if ( Array.isArray( data ) ) {
+		return data.every( ( item ) => typeof item === 'string' || ( typeof item === 'object' && verifyFunction( item ) ) );
+	}
+
+	// Arguably this should fail/throw, because none of our allowed types were encountered.
+	return false;
+}
+
+/**
+ * Verifies that provided metrics match allowed values. Metrics can be a string,
+ * an array of string, an array of objects or mix of strings and objects. Objects
+ * must have "expression" and "alias" properties in order to be considered as valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|string[]|Object|Object[]} metrics The metrics to check.
+ * @return {boolean} TRUE if metrics are valid, otherwise FALSE.
+ */
+export function isValidMetrics( metrics ) {
```

Comment:
Since this is specific to Analytics, I think we should either rename this function to `isValidAnalyticsMetrics` or move it to `assets/js/modules/analytics/util` (in there it could keep the current name since it would have the context from being in that folder).

The alternative here is we make this more generic. This means we would diverge a little bit from the original APIs in certain cases, but on the other hand it would be nice if these APIs would all work in a similar way when called via Site Kit - possibly makes it more trivial. On the other hand, if we diverge, reading the API docs could cause false assumptions on that it works the same way in Site Kit - what do you think @eugene-manuilov?

Reply:
@felixarntz I like the idea to make it more generic, but unfortunately I agree with you that it could cause false assumptions and as a result it might confuse developers who will work with it and expect the same behavior from Site Kit. I'll move it to the analytics module.

Reply:
Updated.